aeron-driver/src/main/c/aeron_driver_sender.c

```diff
@@ -115,18 +115,14 @@ int aeron_driver_sender_init(
     return 0;
 }
 
-void aeron_driver_sender_on_command(void *clientd, void *item)
+static void aeron_driver_sender_on_rb_command_queue(
+    int32_t msg_type_id,
+    const void *message,
+    size_t size,
+    void *clientd)
 {
-    aeron_driver_sender_t *sender = (aeron_driver_sender_t *)clientd;
-    aeron_command_base_t *cmd = (aeron_command_base_t *)item;
-    bool is_delete_cmd = cmd->func == aeron_command_on_delete_cmd;
-
+    aeron_command_base_t *cmd = (aeron_command_base_t *)message;
     cmd->func(clientd, cmd);
-
-    if (!is_delete_cmd)
-    {
-        aeron_driver_conductor_proxy_on_delete_cmd(sender->context->conductor_proxy, cmd);
-    }
```

Comment:
This section seems like a worthwhile change. 

The rest of the change seems nice for tidying up the proxy files but I don't think it truly solves the issue.
Afaict the only reason its harder to reproduce now is that you've effectively increased the size of the command queue massively.
In 1.40.0 it was 128 commands long, in this change its 128KB long, Even the largest command doesn't look to be larger than 100 bytes with the ring buffer header, so the queue is holding a minimum of 1000 commands now.

I see that today the queue size has been increased in master anyway https://github.com/real-logic/aeron/commit/1f7ff46d37699e2a8e62844ee7cab5b0c8dfcb63.

Either way it seems like its still technically possible to end up in this locked queue situation, it just takes a much higher load to get there.

Reply:
> Afaict the only reason its harder to reproduce now is that you've effectively increased the size of the command queue massively.

While this change does increase the buffer space, that is not the primary fix.  Prior to implementing this change, I experimented with turning the queue size back to 128 and replacing the `*_proxy_on_delete_command` messages with an inline free.  I was unable to replicate this issue with that change in place.

Using a ring buffer instead of a queue was mostly down to coming up with an implementation that avoided the need for `alloc`/`free` of the messages between the agents (thereby removing the need for the `*_proxy_on_delete_command` messages altogether).  We didn't want to go with the inline `free` as the fix because (depending on allocator) you can get jitter in the call to `free`, especially if the memory moves between threads.  This is something that we actively try to avoid in the sender/receiver.

As you mention it is still technically possible to get a deadlock situation, but it is not a factor of load.  It requires a very specific operation and a lot of load, to point where I think that it is highly unlikely in practical terms (i.e. to trigger it would probably require artificial load rather than that of a real application).  That's not to say that we won't fix it, but a more general solution requires more thought as it may result in a behaviour change that impacts users.  Even with a general solution in place we wanted to make this change anyway as it simplifies the code.

Reply:
Interesting, that gives me more confidence. I would've liked to try reproducing with the inline free version as well (if you still have this I'm happy to do so now).

Regardless of that, thanks for the full explanation of the change, and the speed of your responses/fix!

Reply:
I don't, but you can try it by replacing the implementation of `aeron_driver_conductor_proxy_on_delete_cmd` with a free instead of sending the message.  We are also planning on some changes to that will back pressure the client as the queues to get full.